_posts/2017-08-16-json-server.md

```diff
@@ -0,0 +1,655 @@
+
+As a matter of fact, this URI is not as intuitive as it could be. It could be argued that the following URI would offer better clarity: `POST http://localhost:3000/articles/1/comments`. Now we know we're creating a comment in article id 1.
+
+```bash
+$ curl --data-urlencode "body=Cool article ;-)" http://localhost:3000/articles/1/comments
+{
+  "body": "Cool article ;-)",
+  "articleId": 1,
+  "id": 4
+}
+```
+
+Same with creating an article by author id 3:
+
+```bash
+$ curl --data-urlencode "title=GraphQL" http://localhost:3000/authors/3/articles
+{
+  "title": "GraphQL",
+  "authorId": "3",
+  "id": 3
+}
+```
+
+<a name="filters"></a>
+
+#### Filters, sort & operators
+
+**Filtering** is done using query parameters: `GET http://localhost:3000/articles?title=GraphQL`.
+
+**Sort** is as easy as adding `_sort` & `_order` (`asc` & `desc`) query parameters: 
+
+`GET http://localhost:3000/articles?_sort=likes`
+
+(Assuming we have added the `likes` field to each article). The order is ascending by default.
+
+In case we want to sort by multiple properties, we can write our properties separated by a comma:
+
+`GET http://localhost:3000/articles?_sort=author,score&_order=desc,asc`
+
+**Operators** are suffixes used to augment our query parameters:
+
+* `_gt` (greater than), `_lt` (less than), `_gte` (greater than or equal) and `_lte` (less than or equal): `GET http://localhost:3000/comments?score_gte=5` (assuming we have a `score` field in the comments)
+* `_ne`(not equal) negation of an expression `GET http://localhost:3000/comments?articleId_ne=2`
+* `_like` is an operator that can be applied to string properties, it gives the same result as an `SQL`'s `LIKE`. `GET http://localhost:3000/articles?title_like=API`
+
+<a name="pagination"></a>
+
+#### Pagination
+
+We can use the built-in query parameters `_page` & `_limit` to paginate our results.
+`json-server` exposes `X-Total-Count` and the `Link` header that contain links to the first, next and last pages.
+
+`GET http://localhost:3000/articles?_page=1&_limit=1`
+
+```http
+HTTP/1.1 200 OK
+X-Powered-By: Express
+Vary: Origin, Accept-Encoding
+Access-Control-Allow-Credentials: true
+Cache-Control: no-cache
+Pragma: no-cache
+Expires: -1
+X-Total-Count: 3
+Access-Control-Expose-Headers: X-Total-Count, Link
+Link: <http://localhost:3000/articles?_page=1&_limit=1>; rel="first", <http://localhost:3000/articles?_page=2&_limit=1>; rel="next", <http://localhost:3000/articles?_page=3&_limit=1>; rel="last"
+X-Content-Type-Options: nosniff
+Content-Type: application/json; charset=utf-8
+Content-Length: 89
+ETag: W/"59-24+hjZrVFdbtnn+FgcogU6QvujI"
+Date: Sun, 30 Jul 2017 17:22:34 GMT
+Connection: keep-alive
+```
+
+<a name="full-text-search"></a>
+
+#### Full text search
+
+We can implement a search feature in our app using full-text search by simply adding a `q` query parameter.
```

Comment:
We can implement in our app a search feature using full-text search by simply adding a `q` query parameter.